avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
         :param expire: time in seconds for the asset to expire
         :param metadata: metadata which will be saved inside metadata file
         """
-        self.name = name
+        parsed_url = urlparse.urlparse(name)
+        # Asset file name
+        self.basename = os.path.basename(parsed_url.path)
         self.asset_hash = asset_hash
-        if algorithm is None:
-            self.algorithm = DEFAULT_HASH_ALGORITHM
-        else:
+
+        if algorithm:
             self.algorithm = algorithm
+        else:
+            self.algorithm = DEFAULT_HASH_ALGORITHM
 
+        self.locations = []
+        # If `name` is an url, move it to `locations`
+        if parsed_url.scheme:
+            self.locations.append(os.path.dirname(parsed_url.geturl()))
+        # Let's be conservative and check `locations` type
         if isinstance(locations, str):
-            self.locations = [locations]
+            # remove asset name from location
+            if self.basename in locations:
+                parsed_locations = urlparse.urlparse(locations)
+                self.locations.append(
+                    os.path.dirname(parsed_locations.geturl()))
+            else:
+                self.locations.append(locations)
+        elif isinstance(locations, list):
+            # remove asset name from location
+            for location in locations:
+                if self.basename in location:
+                    parsed_location = urlparse.urlparse(location)
+                    self.locations.append(
+                        os.path.dirname(parsed_location.geturl()))
+                else:
+                    self.locations.append(location)
+
+        # Let's make the user's life easy and handle the cache
+        if cache_dirs:
+            self.cache_dirs = cache_dirs
         else:
-            self.locations = locations
-        self.cache_dirs = cache_dirs
+            # Break early if we don't have the cache_dirs. This is necessary
+            # so we don't change the Asset signature.
+            raise EnvironmentError("At least one cache directory is necessary"
+                                   " to download the asset.")
+
         self.expire = expire
         self.metadata = metadata
 
+        # Break early. We need to know the source location of the asset.
+        if not self.locations:
+            raise EnvironmentError("An asset should have at least one source"
+                                   " location. Double check your asset name"
+                                   " and location and make sure to use a"
+                                   " supported protocol.")
+
+    def _create_hash_file(self, asset_path):
+        """
+        Compute the hash of the asset file and add it to the CHECKSUM
+        file.
+
+        :param asset_file: full path of the asset file.
+        """
+        result = crypto.hash_file(asset_path, algorithm=self.algorithm)
```

Comment:
Just wondering if `crypto.hash_file` should still be maintained. By the time it was introduced perhaps Avocado supported only Python-2, but now with Python-3 only support it could be replace with `hashlib` module?
@clebergnu time for archaeology  ... :)

Reply:
ah, talking about Python 2 vs 3... now that Avocado supports Python 3 only perhaps we can remove from this file:
```
try:
    import urlparse
except ImportError:
    import urllib.parse as urlparse
```

Reply:
Good catch! Python 2 vs 3 was not something I was checking while working on this PR.

Reply:
> Just wondering if `crypto.hash_file` should still be maintained. By the time it was introduced perhaps Avocado supported only Python-2, but now with Python-3 only support it could be replace with `hashlib` module?
> @clebergnu time for archaeology ... :)

If you look at `crypto.hash_file()` you'll see that it uses `hashlib`, but it's a utility wrapper intended to avoid the pattern:

1. Open file
2. Read at most `n` bytes (if `size` is not `None`)
3. Calculate hash
4. Returns hex representation

So unless this pattern is not found to be useful (and repeatedly used), I think it should still be maintained.